isassigned: check bounds on multidimensional indexes

shlemas opened the pull request:
Fixes #47942

shlemas committed:
isassigned: check bounds on multidimensional indexes

Seelengrab commented on base/array.jl:
```diff
@@ -215,9 +215,9 @@ sizeof(a::Array) = Core.sizeof(a)
-    ii = (_sub2ind(size(a), i...) % UInt) - 1
-    @boundscheck ii < length(a) % UInt || return false
-    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
+    @boundscheck isempty(i) || checkbounds(Bool, a, i...) || return false
```
Why the `isempty(i)`? `checkbounds(Bool, a, i...)` works with empty splats as well and correctly returns `false`, leading to `return false`. `isempty` would return `true`, passing the boundscheck through shortcircuiting and subsequently checking whether the first element is assigned for e.g. an empty vector, which doesn't make sense and would be an OOB again, right?

```
julia> Base._sub2ind(size([]))
1
```

I think just `checkbounds(Bool, a) || return false` should be sufficient.

shlemas replied:
This was to support existing unit tests like:
```
@test isassigned(a)
```

Basically `checkbounds` returns `false` for an empty splat, but `isassigned` returns `true`.

stevengj replied:
Seems like the latter behavior might be another bug in `isassigned`.

shlemas replied:
After looking a little closer tonight, it's inconsistent. For example:

* `isassigned` for `Array` (the one we're talking about) returns `true` for an empty splat.
* `isassigned` for `AbstractArray` returns `false` for an empty splat.

Since they're currently inconsistent anyway, making them both return `false` sounds good. So I tried that and ran into the next issue caused by:
```
julia> checkbounds(Bool, [1])
true

julia> checkbounds(Bool, [1, 2])
false

julia> checkbounds(Bool, rand(1,1,1,1))
true
```

I think this behavior was introduced in https://github.com/JuliaLang/julia/commit/f65c403d7335acaaeb6d5ab16a6b6da679307ae6

It sounds like the intent was to preserve an ability to omit indexes for dimensions of length 1. But then arrays/vectors with one element get caught up in that. Can probably fix with a couple specializations of `checkbounds_indices`, I'll try it tomorrow.

Seelengrab replied:
I think the `checkbounds` behavior for non-empty arrays is intentional. Consider these:

```
julia> [2][]
2

julia> [2;;;;;] |> typeof
Array{Int64, 5}

julia> [2;;;;;][]
2

julia> [2, 1][]
ERROR: BoundsError: attempt to access 2-element Vector{Int64} at index []
Stacktrace:
 [1] throw_boundserror(A::Vector{Int64}, I::Tuple{})
   @ Base ./abstractarray.jl:744
 [2] checkbounds
   @ ./abstractarray.jl:709 [inlined]
 [3] _getindex
   @ ./abstractarray.jl:1326 [inlined]
 [4] getindex(::Vector{Int64})
   @ Base ./abstractarray.jl:1294
 [5] top-level scope
   @ REPL[11]:1
```

Which ought to be entirely correct. So the only inconsistency I can see is for the empty case, where the current code returns `false` and yours would return true:

```
julia> ccall(:jl_array_isassigned, Cint, (Any, UInt), [], 0) == 1
true
```

shlemas replied:
That cleared things up, thanks! Removing the `isempty` check is all that needed to be done after all then. Everything seems to be looking sane and consistent now.

Seelengrab replied:
Yeah, the special casing of singleton things is sometimes a bit tricky to wrap your head around :)

shlemas replied:
One more question for you since I think I misread this last night:

> I think the `checkbounds` behavior for non-empty arrays is a bug. Consider these:

When I run your examples through `checkbounds`, it look consistent with the results you posted:

```
julia> checkbounds(Bool, [2])
true

julia> checkbounds(Bool, [2;;;;;])
true

julia> checkbounds(Bool, [2, 1])
false
```

Is there anything else to look at there?

Seelengrab replied:
No, I miswrote `checkbounds` for `isassigned` - that's why I shouldn't write these comments on a lazy holiday :sweat_smile: `checkbounds` should always agree with whether or not a `BoundsError` is thrown :)

shlemas replied:
I probably misread it as `isassigned` last night anyway because I was up way too late. 😛 That all makes sense. Thank you!

shlemas committed:
isassigned: remove isempty check

This check was added in the previous commit to preserve existing
behavior when calling isassigned with an empty splat for the indexes.
But it turns out different version of isassigned treat this case
differently (e.g. AbstractArray vs Array). So the check is being
removed to make isassigned behavior consistent and more in line
with checkbounds and expectations of the [] operator.

Seelengrab commented on base/array.jl:
```diff
@@ -215,8 +215,8 @@ sizeof(a::Array) = Core.sizeof(a)
 
 function isassigned(a::Array, i::Int...)
```
Yes, though this in particular seems hard to test :thinking: Is there a similar test for `getindex(::Array, ::Int)` we could steal?

shlemas replied:
Started looking at this out of curiosity.

I found `test/boundscheck.jl` which executes `test/boundscheck_exec.jl` with different values for the `--check-bounds` flag. There are tests in there for `@inbounds` etc. for both `isassigned` and `getindex`.

I think something is maybe wrong with running the tests though. This test fails for me on master when I run it manually:
```
./julia --check-bounds=auto --startup-file=no --depwarn=error test/boundscheck_exec.jl
```

Same with the higher level script:
```
cd test/
../julia boundscheck.jl
```

It seems the `--check-bounds=auto` (default) case isn't executing during `make testall` or `make test[-revise]-boundscheck`. It isn't for me at least. I can throw `@test false` into those code paths and it won't hit it. As a result it seems a test failure may have snuck into the repo.

shlemas replied:
M'kay that was happening because `Base.julia_cmd()` returns a command with the same flags as the running Julia process, and all the tests are run with `--check-bounds=yes` by default, so it was inheriting that. Need to add an explicit `--check-bounds=auto` in: https://github.com/JuliaLang/julia/blob/79e29e3cb9ef34f7ab613657c0f480d81b1402b7/test/boundscheck.jl#L5

Then the failing test runs which I guess is a whole other issue:
```
Test Failed at test/boundscheck_exec.jl:35
  Expression: A1_wrap() == 0
   Evaluated: 1 == 0
```

Seelengrab replied:
Yes, that failure seems unrelated (though nonetheless bad), thanks for looking into it! That indeed seems a bit broken. `A1_wrap` wraps `A1_inbounds`, which I'd expect to be `0`, since it has its own internal `@inbounds` that eliminates the boundscheck. Also interesting that `A1_noinline()` does not seem to be tested.

Please open an issue for this, or a PR for the `--check-bounds=auto` change so there's an explicit place for discussion.

shlemas replied:
Made https://github.com/JuliaLang/julia/issues/48099

shlemas replied:
The tests are fixed on master now, so I was looking again at changing `@inline` -> `@propagate_inbounds`. If I understand correctly how the propagation works (big "if"), I don't think it makes any concrete difference at the moment. None of the functions directly called by `isassigned` has any `@boundscheck`. It looks like I'd have to propagate through quite a few more layers before making it to something like `getindex` where maybe it would matter. Thoughts?

Seelengrab replied:
`@boundscheck` in a caller is not required for `@propagate_inbounds` to work, only `@inbounds` is. As an example:

```julia
function foo(arr, e)
      s = 0
      @inbounds for i in 1:e
          s += bar(arr, i)
      end
      s
end

function bar(arr, idx)
       @boundscheck 1 <= idx <= length(arr)
       z = arr[idx]
       z^2 + 20z
end
```

would be something where the `@boundscheck` can be removed, since the caller guarantees via the presence of the `@inbounds` that they checked that all array indices will be inbounds. I think that's where it could make a difference with `@inline` vs. `@propagate_inbounds`, since there's not enough information in `foo` itself for the compiler to figure it out by itself. Indeed, using `@inline` is not enough there:

```julia
julia> function bar(arr, idx)
              @inline
              @boundscheck 1 <= idx <= length(arr)
              z = arr[idx]
              z^2 + 20z
       end
bar (generic function with 1 method)


julia> @code_native foo(Int[], 1)
# [...]
.LBB0_7:                                # %oob
	mov	rsp, rsi
	movabs	rax, offset ijl_bounds_error_ints
	mov	edx, 1
	mov	qword ptr [rsi], r8
	call	rax
.Lfunc_end0:
	.size	julia_foo_707, .Lfunc_end0-julia_foo_707
	.cfi_endproc
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```

shlemas replied:
Thanks, I misunderstood the problem `@propagate_inbounds` was solving. Playing around with your foo/bar example cleared it up.

I don't seem to get the same behavior with `isassigned`. The compiler seems to be happy in this hacky example with `@inline`, the call to `checkbounds` is removed:

```julia
function isassigned2(a::Array, i::Int...)
    @inline
    @boundscheck checkbounds(Bool, a, i...) || return false
    ii = (Base._sub2ind(size(a), i...) % UInt) - 1
    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
end

function isassigned_wrap(a, i)
    r = 0
    @inbounds for ii in 1:i
        r += (isassigned2(a, i) ? ii : 0)
    end
    return r > 0
end

@test isassigned_wrap(Int[], 2)
```

Even still I can submit the one liner change as a PR. Maybe the compiler is fine with this case but would be unhappy without `@propagate_inbounds` in some other case. If I could find one of those cases I could maybe write a test to go along.

Seelengrab replied:
I have found a case where `@propagate_inbounds` matters, but sadly only in auto-vectorizable code https://github.com/eschnett/SIMD.jl/pull/102

I can share the code if you'd like to take a look, but it's a pretty dense function (similar to the one I have in #48236, but written with SIMD.jl)

shlemas replied:
Sure I'll take a look
